server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java

```diff
@@ -101,21 +103,32 @@ protected void doExecute(Task task, SearchShardsRequest searchShardsRequest, Act
             relativeStartNanos,
             System::nanoTime
         );
+
         ClusterState clusterState = clusterService.state();
+        final Map<String, OriginalIndices> groupedIndices = remoteClusterService.groupIndices(
+            original.indicesOptions(),
+            original.indices()
+        );
+        final OriginalIndices originalIndices = groupedIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+        if (groupedIndices.isEmpty() == false) {
+            throw new UnsupportedOperationException("search_shards API doesn't support remote indices " + original);
+        }
+
+        final AtomicReference<Index[]> resolvedLocalIndices = new AtomicReference<>();
+        final Supplier<Index[]> resolvedLocalIndicesSupplier = () -> {
+            resolvedLocalIndices.compareAndSet(
+                null,
+                transportSearchAction.resolveLocalIndices(originalIndices, clusterState, timeProvider)
+            );
+            return resolvedLocalIndices.get();
+        };
+
         Rewriteable.rewriteAndFetch(
             original,
-            searchService.getRewriteContext(timeProvider::absoluteStartMillis),
+            searchService.getRewriteContext(timeProvider::absoluteStartMillis, resolvedLocalIndicesSupplier),
             listener.delegateFailureAndWrap((delegate, searchRequest) -> {
-                Map<String, OriginalIndices> groupedIndices = remoteClusterService.groupIndices(
-                    searchRequest.indicesOptions(),
-                    searchRequest.indices()
-                );
-                OriginalIndices originalIndices = groupedIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
-                if (groupedIndices.isEmpty() == false) {
-                    throw new UnsupportedOperationException("search_shards API doesn't support remote indices " + searchRequest);
-                }
                 // TODO: Move a share stuff out of the TransportSearchAction.
-                Index[] concreteIndices = transportSearchAction.resolveLocalIndices(originalIndices, clusterState, timeProvider);
+                Index[] concreteIndices = resolvedLocalIndicesSupplier.get();
```

Comment:
The `TransportSearchShardsAction` is used only in cross cluster search scenario. My vote would be to ignore this case for now since this is not supported.

Reply:
Ok, i'll add copious comments here to make sure we capture this for later adjustment

Reply:
Done: https://github.com/elastic/elasticsearch/pull/107075/commits/11049f160850a3ed6b75b3ec5d83e1768aecef57